### ISSUE TITLE
freeze extracting cache 

### ISSUE THREAD

**Issue by laoa27** [Cruel/freeShop#202]
hi, freeshop 2.1.3 get freeze in loading page, it starts checking updates and pass good, then begin downloading latest cache: r24 and pass ok too, then appear the message extracting cache and the circle loading gif stop, get totally freeze and appears a 3ds message to restart console

**Comment by Cruel** [comment]
It does this every time you boot it?

**Comment by laoa27** [comment]
yes, well.. since yesterday I cant run freeshop always the same error, I already reinstall it and nothing change my console is an old 3ds and language is set to spanish if it is of some help

**Comment by Cruel** [comment]
Try deleting the cache directory at /3ds/data/freeShop/cache/

**Comment by laoa27** [comment]
Already did it and nothing change same error :-(

**Comment by rsanch1** [comment]
I have the same issue. Though I have the dns settings on my Wii U to keep it from updating but those settings are on the WiiU and not on my router. But if I tether my n3dsxl to my phone and access freeshop, it loads with no problem and I'm able to download and install without any issue.  I also have an o3dsxl that has the same issue.

**Comment by Pobega** [comment]
The issue is that your rate limiting for Github hosted APIs. FreeShop doesn't seem to handle it well, and will download the 'rate limit reached' message and not branch on it (and instead try to extract it) causing a crash.

The file being downloaded is in src/States/SyncState.cpp, specifically `const char *url = "https://api.github.com/repos/Repo3DS/shop-cache/releases/latest";` (line 302 of current master).

@Cruel you could see your current rate limit for this repo using `curl -i https://api.github.com/repos/Repo3DS`.

See this for more info: https://developer.github.com/v3/#rate-limiting

The bootstrap from EncBinUpdater caused Freeload to hang the whole system, resulting in total data loss. I don't have access to a hacked 3DS but looked into the issue for him. I'd recommend catching rate limit JSON and exiting cleanly on that in a future version.

Take care!

**Comment by Pobega** [comment]
@Cruel I just recreated the issue, here is the JSON I am shown:

```json
{
  "message": "API rate limit exceeded for 208.105.51.90. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)",
  "documentation_url": "https://developer.github.com/v3/#rate-limiting"
}
```

Hope this helps!

**Comment by Cruel** [comment]
That I can fix. And I hope that's the issue here because rate limit seems to be ~60 requests per hour and it seems to have persisted longer than that for this user. But maybe it was trying to use the old file or something.

**Comment by Pobega** [comment]
I've noticed that GitHub reports an epoch time where your available requests will reset, but sometimes it stays at 0. I'm not sure what causes this issue, as it seems to be on Github's end. Ive also seen the remaining requests drop from 54 to 0 pretty quickly, and looking through freeShop I couldn't quite tell how often it hits that API but that seemed excessive to me, but could also be related to an issue on Github's end (they've been under DDOS for the past few weeks)

**Comment by Cruel** [comment]
freeShop uses the API twice on boot if you have auto-update enabled. If you use other software that checks Github API you could hit the limit that way too.

Either way, I need to gracefully fail when it's not a valid json or zip archive during auto-update.

**Comment by Pobega** [comment]
@Cruel would you like me to make a separate issue for you with all of this information, since the title of this bug post is a bit ambiguous?

**Comment by DrSalvador555** [comment]
Has anyone figured out how to resolve this? I'm very confused and sadly I don't understand anything about the JSON. I'm a bit of a noob to this stuff. I'm just not sure why all of a sudden it started giving me problems. Not even a complete removal via settings and deleting of SD contents followed by a reinstall fixed it. Quite curious. 

N3DSXL 
SysNAND 9.2
EmuNAND via ReiNAND 11.2

**Comment by Pobega** [comment]
Have you followed the instructions at Github via my link to see if you are
out of API calls? That will let you know if your issue is related to the
issue I brought up.

**Comment by DrSalvador555** [comment]
Alas! As soon as I post and recheck (for the 9th time) I suddenly get in @_@

**Comment by DrSalvador555** [comment]
Strange, I've got the program running but I'm getting a "No title keys found" error instead. Now what? o.o

**Comment by Pobega** [comment]
You may have to wait for Cruel for a response on this, but don't launch
FreeShop from EncTitleKeysUpdater or whatever it's called. That's how my
friend lost all his data.

**Comment by DrSalvador555** [comment]
I don't launch anything like that, just the freeShop program itself. Cruel says in gbatemp "Looking at the error now, not sure yet what it is. Will probably affect anyone auto-updating title keys."

**Comment by DrSalvador555** [comment]
Seems disabling auto-update and manually adding the enctitlekeys.bin doesn't fix the problem, brings us back the first error which restarts the system.

**Comment by DrSalvador555** [comment]
For anyone having this problem, it's been discovered the bug is in r31, so replace the r31 with r30 found at this link (and using WinRAR works fine on it): https://github.com/Repo3DS/shop-cache/releases/download/r30/cache-r30-etc1.tar.gz

Cruel says he'll try to release r32 to fix this. He says:  "Using an older (smaller) title key dump should get it working, I'm thinking it's a memory issue. It loads everything into RAM so there has always been a limit, just never checked to see what it was."
